File: StrukturDataHashing/Hashing.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Random;
import java.util.Scanner;
import java.util.Set;

class HashTable {
    private int size;
    // Menggunakan ArrayList dari ArrayList untuk merepresentasikan Chaining
    private ArrayList<ArrayList<Integer>> table;
    private int count;

    // Konstruktor
    public HashTable(int size) {
        this.size = size;
        this.count = 0;
        this.table = new ArrayList<>(size);
        
        // Inisialisasi setiap slot dengan ArrayList kosong
        for (int i = 0; i < size; i++) {
            this.table.add(new ArrayList<>());
        }
    }

    // Fungsi hash sederhana (Modulo)
    private int hashFunction(int key) {
        return key % size;
    }

    // 1. INPUT DATA
    public boolean insert(int key) {
        int index = hashFunction(key);
        // Pastikan data belum ada di slot tersebut (menghindari duplikasi)
        if (!table.get(index).contains(key)) {
            table.get(index).add(key);
            count++;
            return true;
        }
        return false;
    }

    // 2. HAPUS DATA
    public boolean delete(int key) {
        int index = hashFunction(key);
        // Java membedakan remove(Object) dan remove(index). 
        // Kita bungkus ke Integer object agar menghapus berdasarkan NILAI, bukan indeks list.
        if (table.get(index).contains(key)) {
            table.get(index).remove(Integer.valueOf(key));
            count--;
            return true;
        }
        return false;
    }

    // 3. CARI DATA
    // Mengembalikan array berisi [status_ketemu (1/0), indeks_slot]
    public int[] search(int key) {
        int index = hashFunction(key);
        if (table.get(index).contains(key)) {
            return new int[]{1, index}; // 1 artinya ditemukan
        }
        return new int[]{0, -1}; // 0 artinya tidak ditemukan
    }

    // 4. MENAMPILKAN SEMUA DATA
    public void displayAll() {
        if (count == 0) {
            System.out.println("Hash Table kosong!");
            return;
        }

        System.out.println("\n--- ISI HASH TABLE (Total: " + count + " data) ---");
        for (int i = 0; i < size; i++) {
            // Hanya tampilkan slot yang memiliki isi
            if (!table.get(i).isEmpty()) {
                System.out.print(String.format("Slot [%03d]: ", i));
                for (int j = 0; j < table.get(i).size(); j++) {
                    System.out.print(table.get(i).get(j));
                    if (j < table.get(i).size() - 1) {
                        System.out.print(" -> ");
                    }
                }
                System.out.println();
            }
        }
        System.out.println("------------------------------------------------");
    }

    // Getter untuk melihat jumlah data saat ini
    public int getCount() {
        return this.count;
    }
}

// --- Program Utama ---
public class Hashing {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Random random = new Random();
        
        // Inisialisasi Hash Table dengan ukuran 150 slot
        HashTable hashTable = new HashTable(150);

        // Mengisi 100 data acak unik di awal program menggunakan HashSet
        Set<Integer> angkaRandom = new HashSet<>();
        while (angkaRandom.size() < 100) {
            // Menghasilkan angka acak antara 1 sampai 1000
            angkaRandom.add(random.nextInt(1000) + 1);
        }

        // Masukkan semua data unik tersebut ke dalam Hash Table
        for (int angka : angkaRandom) {
            hashTable.insert(angka);
        }

        System.out.println("=== PROGRAM HASH TABLE (JAVA) ===");
        System.out.println("Berhasil menginput " + hashTable.getCount() + " data acak unik ke dalam Hash Table.");
        System.out.println("---------------------------------");

        // Menu Interaktif
        while (true) {
            System.out.println("\nMENU:");
            System.out.println("1. Input Data");
            System.out.println("2. Hapus Data");
            System.out.println("3. Cari Data");
            System.out.println("4. Tampilkan Semua Data");
            System.out.println("5. Keluar");
            System.out.print("Pilih menu (1-5): ");
            
            String pilihan = scanner.nextLine();

            switch (pilihan) {
                case "1":
                    try {
                        System.out.print("Masukkan angka yang ingin diinput: ");
                        int dataInput = Integer.parseInt(scanner.nextLine());
                        if (hashTable.insert(dataInput)) {
                            System.out.println("Berhasil! Angka " + dataInput + " telah ditambahkan.");
                        } else {
                            System.out.println("Gagal! Angka " + dataInput + " sudah ada di dalam Hash Table.");
                        }
                    } catch (NumberFormatException e) {
                        System.out.println("Input harus berupa angka numerik!");
                    }
                    break;

                case "2":
                    try {
                        System.out.print("Masukkan angka yang ingin dihapus: ");
                        int dataHapus = Integer.parseInt(scanner.nextLine());
                        if (hashTable.delete(dataHapus)) {
                            System.out.println("Berhasil! Angka " + dataHapus + " telah dihapus.");
                        } else {
                            System.out.println("Gagal! Angka " + dataHapus + " tidak ditemukan.");
                        }
                    } catch (NumberFormatException e) {
                        System.out.println("Input harus berupa angka numerik!");
                    }
                    break;

                case "3":
                    try {
                        System.out.print("Masukkan angka yang dicari: ");
                        int dataCari = Integer.parseInt(scanner.nextLine());
                        int[] hasilCari = hashTable.search(dataCari);
                        
                        if (hasilCari[0] == 1) {
                            System.out.println("Angka " + dataCari + " ditemukan di Hash Table pada indeks slot: " + hasilCari[1]);
                        } else {
                            System.out.println("Angka " + dataCari + " TIDAK ditemukan di Hash Table.");
                        }
                    } catch (NumberFormatException e) {
                        System.out.println("Input harus berupa angka numerik!");
                    }
                    break;

                case "4":
                    hashTable.displayAll();
                    break;

                case "5":
                    System.out.println("Terima kasih! Program selesai.");
                    scanner.close();
                    System.exit(0);
                    break;

                default:
                    System.out.println("Pilihan tidak valid. Silakan coba lagi.");
                    break;
            }
        }
    }
}

File: StrukturDataHashing/Hashing.py
import random

class HashTable:
    def __init__(self, size=150):
        self.size = size
        # Membuat hash table dengan slot kosong berbentuk list (Chaining)
        self.table = [[] for _ in range(self.size)]
        self.count = 0

    def _hash_function(self, key):
        """Fungsi hash sederhana menggunakan metode modulo."""
        return key % self.size

    def insert(self, key):
        """1. INPUT DATA"""
        index = self._hash_function(key)
        # Pastikan data belum ada di dalam slot (menghindari duplikasi)
        if key not in self.table[index]:
            self.table[index].append(key)
            self.count += 1
            return True
        return False

    def delete(self, key):
        """2. HAPUS DATA"""
        index = self._hash_function(key)
        if key in self.table[index]:
            self.table[index].remove(key)
            self.count -= 1
            return True
        return False

    def search(self, key):
        """3. CARI DATA"""
        index = self._hash_function(key)
        if key in self.table[index]:
            return True, index
        return False, None

    def display_all(self):
        """4. MENAMPILKAN SEMUA DATA"""
        if self.count == 0:
            print("Hash Table kosong!")
            return

        print(f"\n--- ISI HASH TABLE (Total: {self.count} data) ---")
        for index in range(self.size):
            # Hanya tampilkan indeks yang memiliki isi data
            if self.table[index]:
                # Berguna untuk melihat bagaimana chaining menangani tabrakan data
                format_data = " -> ".join(map(str, self.table[index]))
                print(f"Slot [{index:03d}]: {format_data}")
        print("------------------------------------------------")

# --- Program Utama ---
if __name__ == "__main__":
    # Inisialisasi Hash Table
    hash_table = HashTable(size=150)

    # Mengisi 100 data acak unik di awal program
    angka_random = set()
    while len(angka_random) < 100:
        angka_random.add(random.randint(1, 1000))

    for angka in angka_random:
        hash_table.insert(angka)

    print("=== PROGRAM HASH TABLE ===")
    print(f"Berhasil menginput {hash_table.count} data acak unik ke dalam Hash Table.")
    print("--------------------------")

    # Menu Interaktif
    while True:
        print("\nMENU:")
        print("1. Input Data")
        print("2. Hapus Data")
        print("3. Cari Data")
        print("4. Tampilkan Semua Data")
        print("5. Keluar")
        
        pilihan = input("Pilih menu (1-5): ")

        if pilihan == '1':
            try:
                data = int(input("Masukkan angka yang ingin diinput: "))
                if hash_table.insert(data):
                    print(f"Berhasil! Angka {data} telah ditambahkan.")
                else:
                    print(f"Gagal! Angka {data} sudah ada di dalam Hash Table.")
            except ValueError:
                print("Input harus berupa angka numerik!")

        elif pilihan == '2':
            try:
                data = int(input("Masukkan angka yang ingin dihapus: "))
                if hash_table.delete(data):
                    print(f"Berhasil! Angka {data} telah dihapus.")
                else:
                    print(f"Gagal! Angka {data} tidak ditemukan.")
            except ValueError:
                print("Input harus berupa angka numerik!")

        elif pilihan == '3':
            try:
                data = int(input("Masukkan angka yang dicari: "))
                ketemu, indeks = hash_table.search(data)
                if ketemu:
                    print(f"Angka {data} ditemukan di Hash Table pada indeks slot: {indeks}")
                else:
                    print(f"Angka {data} TIDAK ditemukan di Hash Table.")
            except ValueError:
                print("Input harus berupa angka numerik!")

        elif pilihan == '4':
            hash_table.display_all()

        elif pilihan == '5':
            print("Terima kasih! Program selesai.")
            break
        else:
            print("Pilihan tidak valid. Silakan coba lagi.")